abstract_specification.md

```diff
@@ -233,68 +298,298 @@ Any given storage "engine" upon which SOMA is implemented may have additional fe
+#      2. root asks child_a to close.
+#          3. child_a asks grandchild to close.
+#              4. grandchild closes itself.
+#          5. child_a closes itself.
+#      6. root asks new_child to close.
+#          7. new_child closes itself.
+#      8. root closes itself.
+#
+# external_object remains open because it is not owned by root.
+```
+
+A user may close a child object before closing the parent object.
+Because `close` is idempotent, this is explicitly allowed.
+Closing the child will _not_ otherwise affect the parent object; it will remain open as usual.
+
+### Operation: set
+
+Adds an entry to the collection, overwriting an existing key if one exists already.
+
+```
+set(string key, SOMAObject value, URIType uri_type)
 ```
 
 Parameters:
 
-- uri - location at which to create the object.
-- [platform_config](#platform-specific-configuration) - optional storage-engine specific configuration
+- key - The key to set.
+- value - The value to set the key to. If a user sets a collection entry to a type inconsistent with the type of that key (e.g. a defined key in an Experiment) or with the general type of the collection (e.g. a collection of Measurements), behavior is unspecified. Implementations are encouraged to raise an error if possible, to prevent latent errors when later attempting to use the data.
+- uri_type - How the collection should refer to the URI of the newly-added element, whether by absolute or relative URI. The default is `auto`, which will use a relative URI if possible but otherwise use an absolute URI. If `absolute`, the entry will always use an absolute URI. If `relative`, the entry will always use a relative URI, and an error will be raised if a relative URI cannot be used.
+
+### Operation: add_new\_<var>object_type</var>
+
+Each <code>add_new\_<var>object_type</var></code> method creates a new SOMA dataset in storage, adds it to the collection, and returns it to the user. The newly-created entry has the same `context` value as the existing collection and is [owned by the current collection](#operation-close-collection-types).
+
+```
+add_new_dataframe(string key, string uri = "", ...) -> SOMADataFrame
+add_new_dense_ndarray(string key, string uri = "", ...) -> SOMADenseNDArray
+add_new_sparse_ndarray(string key, string uri = "", ...) -> SOMASparseNDArray
+```
+
+Parameters:
+
+- key: The key to add the new element at. This cannot already be a key of the collection.
+- uri: An optional parameter to specify a URI to create the new collection, which may be [relative or absolute](#collection-entry-uris). If the URI is relative, the new entry will be added with that relative URI. If the URI is absolute, the new entry will be added with that absolute URI. If a collection already exists at the user-provided URI, the operation should fail. If the user does not specify a URI, the collection will generate a new URI for the entry. When possible, this should be a relative URI based on a sanitized version of the key.
```

Comment:
Do this need to mention that relative URIs cannot be "multi-level"? I think that's what the spec mentions above.